use string if it's a string

Alex-Jordan opened the pull request:
I think this fixes the issue reported by @somiaj at #2117.

Note that with testing, you may want to manually open up `simple.conf` to see what's in there after each test, and clear it for a new test.

Once this is good for `develop` we could do a hotfix.

drgrice1 reviewed (approved):
This looks okay.

There is some oddity in the behavior though.  If simple.conf contains `$permissionLevels{receive_feedback} = 'nobody';`, and you save from any config tab except the E-Mail tab, then the string is converted to an array containing that string, so that file contains `$permissionLevels{receive_feedback} = ['nobody'];`.  If you instead save from the E-Mail tab then the current value is essentially dropped, and the emtpy array  is saved, so that the file contains `$permissionLevels{receive_feedback} = [];`.  This should be okay as these values work as expected.  It is just a bit odd.

In looking closely at the `fetchEmailRecipients` method of the `WeBWorK::Utils` module.  This is my fault, and I can put in a separate pull request to fix this.  I moved the addition of the contents of the `$ce->{mail}{feedbackRecipients}` array into the `fetchEmailRecipients` method.  However, that method may return before that is added.  Those recipients should always be added to the list.

somiaj commented:
This issue fixes things in that the course configuration page no longer errors out if someone has an old course with `$permissionLevels{receive_feedback}` set to a string in `simple.conf`.

I also noticed a difference between saving on the email tab vs other tabs. For instance, if `simple.conf` has `$permissionLevels{receive_feedback} = 'professor'`. If saving on the email tab this is correctly updated to `$permissionLevels{receive_feedback} = ['professor', 'admin']`, but if saved on any other tab this is saved to `$permissionLevels{receive_feedback} = ['professor']`.

I am also a little cautious with this fix, since it seems to be a fix just to ensure that the course configuration page doesn't error out if there is a string vs array issue (which is probably good to have). But it seems that there is already logic to turn `$permissionLevels{receive_feedback}` from a string to an array which is applied correctly on the email tab. I wonder if this logic could also just be used on other tabs (or maybe this change could happen in the course environment so when `get_value` is run the correct array is returned), so that there is a smooth transition from the old use of this setting to the new.

One last thing, since we are fixing things, can we fix the help for this setting on the course configuration page. The help still describes the old meaning of this setting, `E-mail feedback from students automatically sent to this permission level and higher`.  The `and higher` is no longer accurate and should also be updated.

pstaabp commented:
I agree with @somiaj about this changing the legacy feedback.   Shouldn't be too hard to adapt the existing code to handle the change on other configuration tabs. 

Also, I think it would be a good idea to update the help page too.

drgrice1 commented on lib/WeBWorK/ConfigObject/permission_checkboxlist.pm:
```diff
@@ -25,7 +25,11 @@ sub display_value ($self, $val) {
+		if (ref $self->get_value($self->{c}->ce) eq 'ARRAY') {
+			return @{ $self->get_value($self->{c}->ce) };
+		} else {
+			return $self->get_value($self->{c}->ce);
```
If you change this line to
```suggestion
			return @{ role_and_above($self->{c}->ce->{userRoles}, $self->get_value($self->{c}->ce)) };
```
it should fix the issue that @somiaj mentioned.  If the simple.conf contains `$permissionLevels{receive_feedback} = 'professor';`, then this will save `$permissionLevels{receive_feedback} = ['professor','nobody','admin'];` if "Save" is clicked on a tab other than the E-Mail tab.  If simple.conf contains `$permissionLevels{receive_feedback} = 'nobody';` then this will save `$permissionLevels{receive_feedback} = ['nobody'];`.  Those should be the equivalent of the previous setting.

somiaj replied:
@drgrice1 Wouldn't this code also be run on the checkbox lists on other configuration tabs which are not permission levels?

somiaj replied:
Oh wait, I guess this check is only for permissions, not general checkboxlist (missed that).

Alex-Jordan commented:
The `doc2` on this includes `'use the "Contact instructor" button on any problem page.'`

Should I change that to `"Email Instructor"`?  Should I try to change it to use  the config variable `$feedback_button_name`? Not sure how easy it is to use config var values within the `defaults.config` help content. I don't see a precedent. Also it may complicate things that this has to go through maketext.

drgrice1 commented:
That would fail to work with maketext.  I think that the `doc2` should not specifically refer to the button text at all.  That is if that is possible.

Alex-Jordan commented:
I updated the help language.

As part of that, I changed `e-mail` to `email`. So then I felt compelled to do an additional commit to change all other `e-mail` to `email`.

somiaj commented:
@Alex-Jordan `e-mail` -> `email` might be a bit too much for a hotfix, since it will require lots of strings to be retranslated. Maybe don't include that commit in the hotfix at the end.

Alex-Jordan commented:
I thought the same. That is why I made it in a separate commit.

pstaabp reviewed (approved):
Looks good now.
